#6  DRUP-810: Add SmartDocs to API Catalog module

cnovak opened the pull request:
Fixes #6 
* SmartDocs external javascript libraries added to module
* Added SmartDocs field formatter and javascript to load spec in session storage for SmartDocs angular app
* Changed default field formatter for OpenAPI spec to be SmartDocs
* TODO: Add tests #9

cnovak committed:
DRUP-638: SmartDocs rendering from Drupal site

cnovak committed:
Merge branch '8.x-1.x' of https://github.com/apigee/apigee-api-catalog-drupal into 8.x-1.x-smartdocs

cnovak committed:
DRUP-778: change logic to get file extension

cnovak committed:
DRUP-810: Add SmartDocs to API Catalog module

Use Google CDN to host files

cnovak committed:
DRUP-810: Use v1.0.1 of CDN files

Files are now minified/optimized from `ng build —prod` command. No longer need `(window as any).global = window;` fix.

cnovak committed:
DRUP-810: use smartDocs as default field formatter

cnovak committed:
DRUP-810: Formatter cleanup

* Make sure formatter works when there are no specs to load
* Remove code to pass OpenAPI spec URl to javascript, it is not used
* Remove view() function, no longer needed

Jaesin commented on apigee_api_catalog.module:
```diff
@@ -40,3 +40,18 @@ function apigee_api_catalog_help($route_name, RouteMatchInterface $route_match)
+/**
+ * Implements hook_page_attachments().
+ */
+function apigee_api_catalog_page_attachments(array &$attachments) {
```
Could this not be added to the formatter instead of `hook_page_attachments`?

cnovak replied:
OK, got it working now, was not able to get formatter to properly add it earlier, was thinking it was not possible.

Jaesin commented on src/Plugin/Field/FieldFormatter/SmartDocsFormatter.php:
```diff
@@ -0,0 +1,163 @@
+    /** @var \Drupal\file\Entity\File $file */
+    foreach ($this->getEntitiesToView($items, $langcode) as $delta => $file) {
+      $file_extension = pathinfo($file->getFilename(), PATHINFO_EXTENSION);
+      $file_data = file_get_contents($file->getFileUri());
```
This could be allot of data to load/parse to add server side validation using allot of memory. How does the client side app handle an invalid spec. It might be better to validate when the file is saved as opposed to every time it is rendered.

cnovak replied:
The Angular app needs a javascript object, so we need to parse the YAML file into JSON for the Angular app to consume.  We could parse the YAML file in the Angular app after adding an appropriate YAML parser to the Angular, but after talking to SmartDocs devs they thought converting it beforehand would be better.  I tried not parsing JSON files and only YAML, but if you pass the raw file to drupalSettings, the file is treated as a string which makes quotes get double encoded when adding to session storage.  Validating valid YAML/JSON would be good during save also however, but even better would be to validate that it is a valid OpenAPI spec.

cnovak replied:
Updated PR to have the field formatter put the file URL and file extension into drupalSettings instead of actual file.  The smartdos_integration.js script is now responsible for pulling file from server and parsing into javascript to save into session storage for SmartDocs Angular app to retrieve.

Jaesin replied:
Awesome!

mxr576 commented on src/Plugin/Field/FieldFormatter/SmartDocsFormatter.php:
```diff
@@ -0,0 +1,163 @@
+   */
+  public function __construct($plugin_id, $plugin_definition, FieldDefinitionInterface $field_definition, array $settings, $label, $view_mode, array $third_party_settings, LoggerChannelFactoryInterface $loggerFactory) {
+    parent::__construct($plugin_id, $plugin_definition, $field_definition, $settings, $label, $view_mode, $third_party_settings);
+    $this->loggerFactory = $loggerFactory->get('apigee_api_catalog');;
```
You should inject the logger and not the factory.

mxr576 commented on src/Plugin/Field/FieldFormatter/SmartDocsFormatter.php:
```diff
@@ -0,0 +1,163 @@
+  protected $loggerFactory;
+
+  /**
+   * Constructs a EntityReferenceEntityFormatter instance.
```
Outdated docblock 🙂

mxr576 commented on src/Plugin/Field/FieldFormatter/SmartDocsFormatter.php:
```diff
@@ -0,0 +1,163 @@
+      $file_data = file_get_contents($file->getFileUri());
+
+      // Decode the file to pass to the javascript file.
+      if ($file_extension == 'json') {
```
Strict comparison

mxr576 replied:
If you inject a Symfony serializer that suppors both formats you can get the decoded value with one `$this->serializer->decode()` and catch any decoding errors once, independently from the format.

cnovak committed:
Add base tag in formatter instead of hook_page_attachments

cnovak committed:
Remove library reference to smartdocs styles.js

The styles.css file takes place of styles.js, so styles.js is no longer needed.

Jaesin commented:
I am seeing the spec being rendered. Nice. I am now seeing the permalink issue where you refresh a page after navigating in the rendered specification and then refresh the browser.

This formatter doesn't seem to be compatible with having multiple specs rendered on the same page. This is what I see if I create a view and render the spec: 

<img width="419" alt="image" src="https://user-images.githubusercontent.com/159693/58350268-3d0ca880-7e1a-11e9-82a8-ac12d65fe741.png">

cnovak commented:
You cannot render multiple specs on one page since the Angular app uses the URL of the page to determine which  spec ID to render.  For example, if you are on http://smartdocs.loc:8080/api/2, an Angular service is expected to retrieve spec ID 2 to pass back to the SmartDocs rendering library.

Jaesin commented:
Understandable. We should probably mark that as a know issue for now. It would be difficult to try to keep views from rendering the field.

cnovak committed:
DRUP-810: Load and parse specs on client side

The field formatter now puts the file URL and file extension into drupalSettings instead of actual file.  The smartdos_integration js file is now responsible for pulling file from server and parsing into javascript to save into session storage for SmartDocs Angular app to retrieve.

cnovak committed:
DRUP-810: Update README for SmartDocs formatter

Jaesin commented:
When I try to load https://github.com/OAI/OpenAPI-Specification/blob/master/examples/v3.0/uspto.yaml, I get this:  

<img width="941" alt="image" src="https://user-images.githubusercontent.com/159693/58364403-b03d0b80-7e68-11e9-9a56-a67427bf0850.png">

Jaesin reviewed (approved):
OK, clearing my browser cache has fixed the rendering issue on Firefox 67.

mxr576 reviewed (commented):
> You cannot render multiple specs on one page since the Angular app uses the URL of the page to determine which spec ID to render. For example, if you are on http://smartdocs.loc:8080/api/2, an Angular service is expected to retrieve spec ID 2 to pass back to the SmartDocs rendering library.

In Drupal 7, one API doc was one content (node). IIRC there were also an issue when somone tried to render multiple Smartdocs nodes with Views on a page, but that has been fixed.
Since this is a field formatter, we can not expect that users of this module won't try to render multiple API documentation from the same node with this field formatter.
+1 for documenting this limitation, +100 to find a way to remove this limitation as soon as we can

jacine commented on src/Plugin/Field/FieldFormatter/SmartDocsFormatter.php:
```diff
@@ -0,0 +1,157 @@
+ *   label = @Translation("SmartDocs"),
+ *   description = @Translation("Formats OpenAPI specs with SmartDocs"),
+ *   field_types = {
+ *     "file"
```
@cnovak I think this is supposed to be `file_link` instead of `file`.

jacine replied:
I changed this locally, which then allowed "SmartDocs" to be selected as the formatter, but it also caused a fatal error when trying to view the details page:

```
TypeError: Argument 1 passed to Drupal\Core\Field\Plugin\Field\FieldFormatter\EntityReferenceFormatterBase::getEntitiesToView() must implement interface Drupal\Core\Field\EntityReferenceFieldItemListInterface, instance of Drupal\Core\Field\FieldItemList given, called in /Users/jacine/Sites/kickstart_m10n/web/modules/contrib/apigee_api_catalog/src/Plugin/Field/FieldFormatter/SmartDocsFormatter.php on line 126 in Drupal\Core\Field\Plugin\Field\FieldFormatter\EntityReferenceFormatterBase->getEntitiesToView() (line 42 of /Users/jacine/Sites/kickstart_m10n/web/core/lib/Drupal/Core/Field/Plugin/Field/FieldFormatter/EntityReferenceFormatterBase.php) #0 /Users/jacine/Sites/kickstart_m10n/web/modules/contrib/apigee_api_catalog/src/Plugin/Field/FieldFormatter/SmartDocsFormatter.php(126): Drupal\Core\Field\Plugin\Field\FieldFormatter\EntityReferenceFormatterBase->getEntitiesToView(Object(Drupal\Core\Field\FieldItemList), 'en') #1 /Users/jacine/Sites/kickstart_m10n/web/core/lib/Drupal/Core/Field/FormatterBase.php(80): Drupal\apigee_api_catalog\Plugin\Field\FieldFormatter\SmartDocsFormatter->viewElements(Object(Drupal\Core\Field\FieldItemList), 'en') #2 /Users/jacine/Sites/kickstart_m10n/web/core/lib/Drupal/Core/Field/Plugin/Field/FieldFormatter/EntityReferenceFormatterBase.php(76): Drupal\Core\Field\FormatterBase->view(Object(Drupal\Core\Field\FieldItemList), 'en') #3 /Users/jacine/Sites/kickstart_m10n/web/modules/contrib/apigee_api_catalog/src/Plugin/Field/FieldFormatter/SmartDocsFormatter.php(98): Drupal\Core\Field\Plugin\Field\FieldFormatter\EntityReferenceFormatterBase->view(Object(Drupal\Core\Field\FieldItemList), NULL) #4 /Users/jacine/Sites/kickstart_m10n/web/core/lib/Drupal/Core/Entity/Entity/EntityViewDisplay.php(263): Drupal\apigee_api_catalog\Plugin\Field\FieldFormatter\SmartDocsFormatter->view(Object(Drupal\Core\Field\FieldItemList), NULL) #5 /Users/jacine/Sites/kickstart_m10n/web/core/lib/Drupal/Core/Entity/EntityViewBuilder.php(343): Drupal\Core\Entity\Entity\EntityViewDisplay->buildMultiple(Array) #6 /Users/jacine/Sites/kickstart_m10n/web/core/lib/Drupal/Core/Entity/EntityViewBuilder.php(285): Drupal\Core\Entity\EntityViewBuilder->buildComponents(Array, Array, Array, 'full') #7 /Users/jacine/Sites/kickstart_m10n/web/core/lib/Drupal/Core/Entity/EntityViewBuilder.php(242): Drupal\Core\Entity\EntityViewBuilder->buildMultiple(Array) #8 [internal function]: Drupal\Core\Entity\EntityViewBuilder->build(Array) #9 /Users/jacine/Sites/kickstart_m10n/web/core/lib/Drupal/Core/Render/Renderer.php(378): call_user_func(Array, Array) #10 /Users/jacine/Sites/kickstart_m10n/web/core/lib/Drupal/Core/Render/Renderer.php(195): Drupal\Core\Render\Renderer->doRender(Array, false) #11 /Users/jacine/Sites/kickstart_m10n/web/core/lib/Drupal/Core/Render/MainContent/HtmlRenderer.php(226): Drupal\Core\Render\Renderer->render(Array, false) #12 /Users/jacine/Sites/kickstart_m10n/web/core/lib/Drupal/Core/Render/Renderer.php(582): Drupal\Core\Render\MainContent\HtmlRenderer->Drupal\Core\Render\MainContent\{closure}() #13 /Users/jacine/Sites/kickstart_m10n/web/core/lib/Drupal/Core/Render/MainContent/HtmlRenderer.php(227): Drupal\Core\Render\Renderer->executeInRenderContext(Object(Drupal\Core\Render\RenderContext), Object(Closure)) #14 /Users/jacine/Sites/kickstart_m10n/web/core/lib/Drupal/Core/Render/MainContent/HtmlRenderer.php(117): Drupal\Core\Render\MainContent\HtmlRenderer->prepare(Array, Object(Symfony\Component\HttpFoundation\Request), Object(Drupal\Core\Routing\CurrentRouteMatch)) #15 /Users/jacine/Sites/kickstart_m10n/web/core/lib/Drupal/Core/EventSubscriber/MainContentViewSubscriber.php(90): Drupal\Core\Render\MainContent\HtmlRenderer->renderResponse(Array, Object(Symfony\Component\HttpFoundation\Request), Object(Drupal\Core\Routing\CurrentRouteMatch)) #16 [internal function]: Drupal\Core\EventSubscriber\MainContentViewSubscriber->onViewRenderArray(Object(Symfony\Component\HttpKernel\Event\GetResponseForControllerResultEvent), 'kernel.view', Object(Drupal\Component\EventDispatcher\ContainerAwareEventDispatcher)) #17 /Users/jacine/Sites/kickstart_m10n/web/core/lib/Drupal/Component/EventDispatcher/ContainerAwareEventDispatcher.php(111): call_user_func(Array, Object(Symfony\Component\HttpKernel\Event\GetResponseForControllerResultEvent), 'kernel.view', Object(Drupal\Component\EventDispatcher\ContainerAwareEventDispatcher)) #18 /Users/jacine/Sites/kickstart_m10n/vendor/symfony/http-kernel/HttpKernel.php(156): Drupal\Component\EventDispatcher\ContainerAwareEventDispatcher->dispatch('kernel.view', Object(Symfony\Component\HttpKernel\Event\GetResponseForControllerResultEvent)) #19 /Users/jacine/Sites/kickstart_m10n/vendor/symfony/http-kernel/HttpKernel.php(68): Symfony\Component\HttpKernel\HttpKernel->handleRaw(Object(Symfony\Component\HttpFoundation\Request), 1) #20 /Users/jacine/Sites/kickstart_m10n/web/core/lib/Drupal/Core/StackMiddleware/Session.php(57): Symfony\Component\HttpKernel\HttpKernel->handle(Object(Symfony\Component\HttpFoundation\Request), 1, true) #21 /Users/jacine/Sites/kickstart_m10n/web/core/lib/Drupal/Core/StackMiddleware/KernelPreHandle.php(47): Drupal\Core\StackMiddleware\Session->handle(Object(Symfony\Component\HttpFoundation\Request), 1, true) #22 /Users/jacine/Sites/kickstart_m10n/web/core/modules/page_cache/src/StackMiddleware/PageCache.php(106): Drupal\Core\StackMiddleware\KernelPreHandle->handle(Object(Symfony\Component\HttpFoundation\Request), 1, true) #23 /Users/jacine/Sites/kickstart_m10n/web/core/modules/page_cache/src/StackMiddleware/PageCache.php(85): Drupal\page_cache\StackMiddleware\PageCache->pass(Object(Symfony\Component\HttpFoundation\Request), 1, true) #24 /Users/jacine/Sites/kickstart_m10n/web/core/lib/Drupal/Core/StackMiddleware/ReverseProxyMiddleware.php(47): Drupal\page_cache\StackMiddleware\PageCache->handle(Object(Symfony\Component\HttpFoundation\Request), 1, true) #25 /Users/jacine/Sites/kickstart_m10n/web/core/lib/Drupal/Core/StackMiddleware/NegotiationMiddleware.php(52): Drupal\Core\StackMiddleware\ReverseProxyMiddleware->handle(Object(Symfony\Component\HttpFoundation\Request), 1, true) #26 /Users/jacine/Sites/kickstart_m10n/vendor/stack/builder/src/Stack/StackedHttpKernel.php(23): Drupal\Core\StackMiddleware\NegotiationMiddleware->handle(Object(Symfony\Component\HttpFoundation\Request), 1, true) #27 /Users/jacine/Sites/kickstart_m10n/web/core/lib/Drupal/Core/DrupalKernel.php(693): Stack\StackedHttpKernel->handle(Object(Symfony\Component\HttpFoundation\Request), 1, true) #28 /Users/jacine/Sites/kickstart_m10n/web/index.php(19): Drupal\Core\DrupalKernel->handle(Object(Symfony\Component\HttpFoundation\Request)) #29 {main}.
```
